streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java

```diff
@@ -310,6 +317,30 @@ public void run() {
                     cache.resize(size);
                 }
                 stateConsumer.pollAndUpdate();
+
+                if (fetchDeadline != -1) {
+                    if (fetchDeadline >= time.milliseconds()) {
+                        try {
+                            // we pass in a timeout of zero, to just trigger the "get instance id" background RPC,
+                            // we don't want to block the global thread that can do useful work in the meantime
+                            globalConsumerClientInstanceId = globalConsumer.clientInstanceId(Duration.ZERO);
+                            clientInstanceIdFuture.complete(globalConsumerClientInstanceId);
+                            fetchDeadline = -1;
+                        } catch (final IllegalStateException disabledError) {
+                            clientInstanceIdFuture.complete(null);
+                            fetchDeadline = -1;
+                        } catch (final TimeoutException swallow) {
```

Comment:
I'm a bit concerned if multiple calls to `KafkaStreams#clientInstanceIds()` happen in a row within the time out. I think there is a chance of the earlier one left hanging depending on on when the fetchDeadline is reset.
Actually, since the `clientInstanceIdFuture` is overwritten I think it can't be completed if a second call is made before the first future finishes. 

I think we might need a set of futures that are all compeleted then cleared here